### PR TITLE
Add a redirect handler to open non-notebook files from the cli

### PR DESCRIPTION
**Pull request by jtpio**
This makes it possible to open non-notebook files from the cli, for example:

```bash
jupyter nbclassic setup.py
```

https://user-images.githubusercontent.com/591645/108713912-27f60980-7519-11eb-90e7-d70ba10e2c2e.mp4

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -55,7 +55,7 @@
     python_requires='>=3.6',
     include_package_data=True,
     install_requires = [
-        'jupyter_server~=1.1',
+        'jupyter_server~=1.4',
```
Comment (jtpio) [inline setup.py:58]: I think this bump is needed for https://github.com/jupyter-server/jupyter_server/pull/415

File: nbclassic/notebookapp.py [hunk 2]
```diff
@@ -218,6 +219,17 @@ def initialize_handlers(self):
         """Load the (URL pattern, handler) tuples for each component."""
         # Order matters. The first handler to match the URL will handle the request.
         handlers = []
+
+        # Add a redirect from /notebooks to /edit
+        # for opening non-ipynb files in edit mode.
+        handlers.append(
+            (
+                rf"/({self.file_url_prefix})/(.*)\.(((?!ipynb)).*)",
```
Comment (jtpio) [inline nbclassic/notebookapp.py:227]: This regex might need some tweaking to also handle files without a file extension, for example:

```bash
jupyter nbclassic LICENSE
```

Comment (blink1073) [reply nbclassic/notebookapp.py:227]: Are you waiting on this or is the PR ready for merge?

Comment (jtpio) [reply nbclassic/notebookapp.py:227]: It would be fine to merge as is, and we can open a new issue to track this.

File: setup.py (hunk 1 above)
Comment (Zsailer) [reply setup.py:58]: :+1:

**Review by Zsailer — APPROVED** [review]

**Comment by jtpio** [comment]
Starting `nbclassic` with this PR and:

```bash
jupyter nbclassic LICENSE
```

Throws the following exception in `jupyter_server`:

```
$ jupyter nbclassic LICENSE
Traceback (most recent call last):
  File "/path/to/miniforge3/envs/nbclassic/bin/jupyter-nbclassic", line 33, in <module>
    sys.exit(load_entry_point('nbclassic', 'console_scripts', 'jupyter-nbclassic')())
  File "/path/to/miniforge3/envs/nbclassic/lib/python3.9/site-packages/jupyter_server/extension/application.py", line 500, in launch_instance
    subapp = _preparse_for_subcommand(cls, args)
  File "/path/to/miniforge3/envs/nbclassic/lib/python3.9/site-packages/jupyter_server/extension/application.py", line 42, in _preparse_for_subcommand
    if re.match(r'^\w(\-?\w)*$', subc) and subc in Application.subcommands:
TypeError: argument of type 'Dict' is not iterable
```

It looks like passing a file name without an extension like `LICENSE` makes the regex check truthy here and reveals the issue:

https://github.com/jupyter-server/jupyter_server/blob/9b9b311feddf1a6769ad16553357a87984e98085/jupyter_server/extension/application.py#L42

File: nbclassic/notebookapp.py [hunk 3]
```diff
@@ -126,6 +127,7 @@ class NotebookApp(
     aliases = aliases
     flags = flags
     extension_url = "/tree"
+    subcommands = {}
```
Comment (jtpio) [inline nbclassic/notebookapp.py:130]: This should fix the issue mentioned in https://github.com/jupyterlab/nbclassic/pull/45#issuecomment-784233906.

Although it's possible an upstream fix `jupyter_server` could also work.

**Comment by jtpio** [comment]
Updated the regex to also handle files without an extension.

Tested locally with:

- `LICENSE`
- `setup.py`
- `Untitled.ipynb`

File: nbclassic/notebookapp.py (hunk 3 above)
Comment (Zsailer) [reply nbclassic/notebookapp.py:130]: :+1:

**Review by Zsailer — APPROVED** [review]

**Comment by Zsailer** [comment]
Thanks, @jtpio. I verified this works locally for me too.

**Pull request merged**